### PR TITLE
Review import

### PR DESCRIPTION
**Pull request by anhtrantuan**
## TODO checklist

- [x] Implement import service
- [x] Add AJAX to import form
- [x] Write unit tests
- [ ] Update UI
- [ ] Write integration tests

### CODE REVIEW

**Comment by anhtrantuan** [comment]
@damhonglinh This is an interesting challenge! Please let me finish all the tasks before you review! Thanks! 😄

**Comment by anhtrantuan** [comment]
Please note that here are some changes that are unrelated to import feature but still kept in this same PR for simplicity:
- Setup Overcommit hooks.
- Setup test environment.
- Refactoring: Extract validators to a separate class.
- UI enhancement.

File: Gemfile [hunk 1]
```diff
@@ -24,6 +24,7 @@ gem 'jbuilder', '~> 2.0'
 gem 'sdoc', '~> 0.4.0', group: :doc
 
 gem 'haml'
+gem 'remotipart'
```
Comment (anhtrantuan) [inline Gemfile]: To support AJAX file upload.

File: Gemfile [hunk 2]
```diff
@@ -38,14 +39,23 @@ group :development, :test do
...
+  gem 'factory_bot_rails'
   gem 'pry-byebug'
+  gem 'pry-rails'
+  gem 'quiet_assets'
+  gem 'rspec-rails', '~> 3.5'
+end
+
+group :development do
+  # Access an IRB console on exception pages or by using <%= console %> in views
+  gem 'web-console', '~> 2.0'
```
Comment (anhtrantuan) [inline Gemfile]: Move it here to avoid complaint that web-console is activated in test environment.

File: app/services/review_services/import.rb [hunk 3]
```diff
@@ -1,8 +1,40 @@
+require 'csv'
```
Comment (anhtrantuan) [inline app/services/review_services/import.rb]: Ridiculously, it always complains that `CSV` is undefined though it's a built-in Ruby module.
Any suggestion? 🤔 
Otherwise, I'll keep this workaround in intact until there is a solution.

File: app/services/review_services/import_validator.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    private
+
+    attr_reader :file, :encoding
+
+    def validate_file_encoding
+      file_content = File.open(file.path, 'r', encoding: encoding).read
+      return if file_content.valid_encoding?
+
+      errors << 'File cannot be encoded with provided encoding'
```
Comment (anhtrantuan) [inline app/services/review_services/import_validator.rb]: Let's ignore i18n for now! 😄

File: spec/services/review_services/import_validator_spec.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  context 'when file and encoding are valid' do
+    Given(:encoding) { 'MacRoman' }
+    Then { validator.validate }
+  end
+
+  context 'when file does not match encoding' do
+    Given(:encoding) { 'UTF-8' }
+    Then { !validator.validate }
+    And { errors.include?('File cannot be encoded with provided encoding') }
```
Comment (anhtrantuan) [inline spec/services/review_services/import_validator_spec.rb]: Debt incurred from missing i18n. 😄 
https://github.com/judgeme/interview-challenge-tuan-anh/pull/1/files#r264349744

File: Gemfile [hunk 6]
```diff
@@ -38,14 +39,25 @@ group :development, :test do
...
+end
+
+group :test do
+  gem 'capybara'
+  gem 'database_cleaner'
+  gem 'faker'
+  gem 'rspec-given'
+  gem 'selenium-webdriver'
```
Comment (anhtrantuan) [inline Gemfile]: Required by Capybara when running AJAX specs. `geckodriver` binary is required.

File: spec/support/wait_for_ajax.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+module WaitForAjax
```
Comment (anhtrantuan) [inline spec/support/wait_for_ajax.rb]: Reproduced from https://thoughtbot.com/blog/automatically-wait-for-ajax-with-capybara.

**Comment by damhonglinh** [comment]
How many hours have you spent on this PR? Because based on the changes here (not on the timestamps of your commit), it seems to me that you have spent a lot of time on this PR. 

About the changes added `for simplicity` (for overcommit, test set up, etc.),  I'm not sure for whose simplicity because I have to mentally separate those changes from the actual changes needed for the requirements 😄

**Comment by anhtrantuan** [comment]
Yes, I spent quite a lot time on this PR. As I mentioned previously, I would like to make it perfect. It would take me ~3 hours, however, to come up with only basic implementation.

Regarding simplicity, I mean for the process of reviewing in general, not for individual review. Yet, I  2 solutions in case you feel it difficult to review:
1. Separate unrelated changes to another PR. Them base this PR on corresponding branch.
2. Merge unrelated changes directly to master and keep this PR intact.
Please let me know which way you prefer! Thanks! 😄

**Comment by anhtrantuan** [comment]
@damhonglinh @betelgeus13 What's your opinion on my previous comment?
Please let me know soon so that I can finalize this! Thanks! 🙇

**Comment by damhonglinh** [comment]
Yeah, I can see you put a great deal of effort on this PR. Although we don't encourage this but we appreciate your effort! 
If it doesn't cost you much time, can you do `1`, separate the extra changes from the needed changes somehow (probably by modifying this PR to base on another branch, which contains only the extra changes)? But if it costs you some time, just skip it.

**Comment by anhtrantuan** [comment]
@damhonglinh As requested, I have added another PR (https://github.com/judgeme/interview-challenge-tuan-anh/pull/2) and base this PR on that. Please have a look! 😄

File: app/services/review_services/import.rb (hunk 3 above)
Comment (damhonglinh) [reply app/services/review_services/import.rb]: Actually I think `csv` is a standard library but not a built-in module. I guess some std libraries (if not all) have to be required explicitly although I'm not really sure why.

File: app/services/review_services/import.rb [hunk 8]
```diff
@@ -1,8 +1,40 @@
...
+    }.freeze
 
     def initialize(file, encoding)
+      @file = file
+      @encoding = encoding
+      validate_file_and_encoding
+    end
+
+    def import
+      ActiveRecord::Base.transaction do
```
Comment (damhonglinh) [inline app/services/review_services/import.rb]: It depends on the business logics but usually in normal cases, a failed iteration of a loop should fail the whole loop unless the loop does some critical actions (such as bank transactions). So in this case of importing reviews, it's probably better to import as many reviews as possible so it's probably better to ignore failed reviews.

You may argue users may want to fix the CSV file and import the file again, which would create duplicated reviews. It's a valid argument. But it's probably better to have duplicate detection to prevent duplications. But this is out of the scope of this challenge.

File: app/services/review_services/import_validator.rb [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    def validate_csv_row(row, index)
+      fields = row.fields
+      return if fields.size == Import::KEYS.size && fields[1].present?
+
+      errors << "Invalid CSV content at line #{index}"
+    end
+
+    def validate_csv_schema
+      CSV.foreach(
+        file.path,
```
Comment (damhonglinh) [inline app/services/review_services/import_validator.rb]: First time see this style of new line for every arguments. It seems long to me but it's an excellent style for those who bill clients by lines of code. 😆

File: spec/controllers/reviews_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
+require 'rails_helper'
```
Comment (damhonglinh) [inline spec/controllers/reviews_controller_spec.rb]: Pretty long tests for a simple feature and for 4 hours. I appreciate your effort but it seems quite overkill again.

File: Gemfile (hunk 1 above)
Comment (damhonglinh) [reply Gemfile]: Normally I would write my own JS for file upload, it's just as simple as a few more lines (you can see in this codebase). `remotipart` probably has more features but they are not necessary for now.

File: app/services/review_services/import_validator.rb (hunk 4 above)
Comment (damhonglinh) [reply app/services/review_services/import_validator.rb]: Technically, it's "File cannot be **decoded**" but not "encoded". But users probably don't understand it anyway. 😄

File: Gemfile (hunk 1 above)
Comment (anhtrantuan) [reply Gemfile]: Actually, it solves a few problems caused by Turbolinks. But let me try to write my own JS then! 😄

File: app/services/review_services/import_validator.rb (hunk 4 above)
Comment (anhtrantuan) [reply app/services/review_services/import_validator.rb]: Yup, `encoded` should be replaced with `decoded` to be precise. Even better, the error could be `Invalid encoding`.

File: app/services/review_services/import_validator.rb [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+    def validate_file_encoding
+      file_content = File.open(file.path, 'r', encoding: encoding).read
+      return if file_content.valid_encoding?
+
+      errors << 'File cannot be encoded with provided encoding'
+    end
+
+    def validate_csv_header(row)
+      return if row.headers == Import::KEYS.values
```
[the remark this answers is not in the document]
Comment (anhtrantuan) [reply app/services/review_services/import_validator.rb]: It depends on how strict we want the validations to be. I can loose it and ignore column order as suggested.

File: app/controllers/reviews_controller.rb [hunk 12]
```diff
@@ -1,12 +1,21 @@
...
+  # :reek:DuplicateMethodCall
   def import
-    # binding.pry # use this for debugger
+    ReviewServices::Import
+      .new(params.require(:file), params.require(:encoding))
+      .import
+    @reviews = Review.new_to_old
+  rescue ReviewServices::Import::ValidationError => error
+    @error = error.message
+  rescue StandardError => error
```
[the remark this answers is not in the document]
Comment (anhtrantuan) [reply app/controllers/reviews_controller.rb]: Since we use AJAX, this is necessary or users have to open browser dev tool to see the error log. 😅

File: app/services/review_services/import.rb [hunk 13]
```diff
@@ -1,8 +1,40 @@
...
 
+    private
+
+    attr_reader :file, :encoding
+
+    def validate_file_and_encoding
+      validator = ImportValidator.new(file, encoding)
+      return if validator.validate
+
+      raise ValidationError, validator.errors.to_sentence
```
[the remark this answers is not in the document]
Comment (anhtrantuan) [reply app/services/review_services/import.rb]: There's an infinite contro regarding condition- vs exception-based control flow. My last company prefers condition-based control flow while my current company prefers exception-based control flow. Either way is fine to me as long as it's agreed by the team. 😄 

I'll change it to if-else as suggested! 😉

File: app/services/review_services/import.rb (hunk 8 above)
Comment (anhtrantuan) [reply app/services/review_services/import.rb]: Okay, let's remove transaction block here.

File: app/services/review_services/import_validator.rb (hunk 9 above)
Comment (anhtrantuan) [reply app/services/review_services/import_validator.rb]: I split them out into lines to prevent violating the convention of 80-column line. There are a few choices but this looks most comfortable to me. 😄 
```ruby
# Style 1
CSV.foreach(file.path, headers: true, encoding: encoding)
   .with_index(2) do |row, index|
     validate_csv_header(row) if index == 2
     validate_csv_row(row, index)
   end

# Style 2 (my preference)
CSV.foreach(
  file.path,
  headers: true,
  encoding: encoding
).with_index(2) do |row, index|
  validate_csv_header(row) if index == 2
  validate_csv_row(row, index)
end

# Style 3
CSV.foreach(
  file.path, headers: true, encoding: encoding
).with_index(2) do |row, index|
  validate_csv_header(row) if index == 2
  validate_csv_row(row, index)
end

# Style 4
CSV
  .foreach(file.path, headers: true, encoding: encoding)
  .with_index(2) do |row, index|
    validate_csv_header(row) if index == 2
    validate_csv_row(row, index)
  end
```

Btw, I'd charge clients by quality and efforts, not LoC and hours. 😜

File: app/services/review_services/import_validator.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (damhonglinh) [reply app/services/review_services/import_validator.rb]: It's fine, just leave it.

File: spec/controllers/reviews_controller_spec.rb (hunk 10 above)
Comment (anhtrantuan) [reply spec/controllers/reviews_controller_spec.rb]: To be honest, I'm not bound to the idea that this challenge is time-framed. I could have spent only half time to produce a solution. But I would not be happy with that coz I don't "code for food". 😂 
In real work, unless there's something more important to do, I'd not trade off quality. I do understand that time-to-market matters but things have to be balanced or debt will incur. Thus, I'll not fit your company if you don't believe in quality.

File: app/services/review_services/import.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (damhonglinh) [reply app/services/review_services/import.rb]: Yeah, it depends on each team indeed. But it's fine, you can just leave it.

File: app/services/review_services/import_validator.rb (hunk 9 above)
Comment (damhonglinh) [reply app/services/review_services/import_validator.rb]: Yeah, I was just kidding 😉
